format/postgres/common/pg_btree/postgres/pg_btree.go

```diff
@@ -0,0 +1,284 @@
+	})
+	d.FieldStruct("meta_page_data", func(d *decode.D) {
+		decodeBTMetaPageData(d)
+	})
+
+	pos0 := d.Pos()
+	pos1 := btree.page.BytesPosSpecial * 8
+	d.FieldRawLen("unused0", pos1-pos0)
+	d.FieldStruct("page_opaque_data", func(d *decode.D) {
+		decodeBTPageOpaqueData(d)
+	})
+	pos2 := d.Pos()
+	bytesPos2 := pos2 / 8
+	if bytesPos2 != page.BytesPosEnd {
+		d.Fatalf("invalid pos after read page_opaque_data on meta page\n")
+	}
+}
+
+func decodeBTMetaPageData(d *decode.D) {
+	/*    0      |     4 */ // uint32 btm_magic
+	/*    4      |     4 */ // uint32 btm_version
+	/*    8      |     4 */ // BlockNumber btm_root
+	/*   12      |     4 */ // uint32 btm_level
+	/*   16      |     4 */ // BlockNumber btm_fastroot
+	/*   20      |     4 */ // uint32 btm_fastlevel
+	/*   24      |     4 */ // uint32 btm_last_cleanup_num_delpages
+	/* XXX  4-byte hole */
+	/*   32      |     8 */ // float8 btm_last_cleanup_num_heap_tuples
+	/*   40      |     1 */ // _Bool btm_allequalimage
+	/* XXX  7-byte padding */
+
+	btmMagic := d.FieldU32("btm_magic")
+	d.FieldU32("btm_version")
+	d.FieldU32("btm_root")
+	d.FieldU32("btm_level")
+	d.FieldU32("btm_fastroot")
+	d.FieldU32("btm_fastlevel")
+	d.FieldU32("btm_last_cleanup_num_delpages")
+	d.FieldU32("hole0")
+	d.FieldF64("btm_last_cleanup_num_heap_tuples")
+	d.FieldU8("btm_allequalimage")
+	d.FieldU56("padding0")
+
+	if btmMagic != BTREE_MAGIC {
+		d.Fatalf("invalid btmMagic = %X, must be %X\n", btmMagic, BTREE_MAGIC)
+	}
+}
+
+// struct BTPageOpaqueData {
+/*    0      |     4 */ // BlockNumber btpo_prev;
+/*    4      |     4 */ // BlockNumber btpo_next;
+/*    8      |     4 */ // uint32 btpo_level;
+/*   12      |     2 */ // uint16 btpo_flags;
+/*   14      |     2 */ // BTCycleId btpo_cycleid;
+func decodeBTPageOpaqueData(d *decode.D) {
+	prev := d.FieldU32("btpo_prev")
+	next := d.FieldU32("btpo_next")
+	d.FieldU32("btpo_level")
+	flags := d.FieldU16("btpo_flags")
+	d.FieldU16("btpo_cycleid")
+
+	isLeftMost := prev == P_NONE
+	isRightMost := next == P_NONE
+	isLeaf := (flags & BTP_LEAF) != 0
+	isRoot := (flags & BTP_ROOT) != 0
+	isDeleted := (flags & BTP_DELETED) != 0
+	isMeta := (flags & BTP_META) != 0
+	isHalfDead := (flags & BTP_HALF_DEAD) != 0
+	isIgnore := isDeleted || isHalfDead
+	hasGarbage := (flags & BTP_HAS_GARBAGE) != 0
+	isIncompleteSplit := (flags & BTP_INCOMPLETE_SPLIT) != 0
+	hasFullXid := (flags & BTP_HAS_FULLXID) != 0
+
+	d.FieldStruct("flags", func(d *decode.D) {
+		d.FieldValueBool("is_leftmost", isLeftMost)
+		d.FieldValueBool("is_rightmost", isRightMost)
+		d.FieldValueBool("is_leaf", isLeaf)
+		d.FieldValueBool("is_root", isRoot)
+		d.FieldValueBool("is_deleted", isDeleted)
+		d.FieldValueBool("is_meta", isMeta)
+		d.FieldValueBool("is_half_dead", isHalfDead)
+		d.FieldValueBool("is_ignore", isIgnore)
+		d.FieldValueBool("has_garbage", hasGarbage)
+		d.FieldValueBool("is_incomplete_split", isIncompleteSplit)
+		d.FieldValueBool("has_full_xid", hasFullXid)
+	})
```

Comment:
Think i commend about this before but don't remember what the conclusion was: Would it make sense to decode so that there will be fields for each "real" bit or it does not make sense? i look like most flags are mapped directly to one bit but some are "derived" like isIgnore

Reply:
It useful when analysing data such as table rows or indexes. I am searching for anomalies and check every flag. I am not so good in applying masks in mind. I better understand boolean in text.

Reply:
Either way works for me, think i let you decide what is best, and we can always change it later if tracking bit rangres per flag is useful etc.

I've thought about adding some kind of helper to the decode API to make it more convenient to produce FieldBool and FieldU/S etc from an endian "aware" bitstream decoder somehow, maybe something similar to how kaitai struct does "bit-endian" (https://doc.kaitai.io/user_guide.html#_bit_sized_integers see "4.14.2. Little-endian order").

But generally i'm fine with having the decode code being a bit ugly or weird as long it produces a nicer decode output to the user but maybe that is not the case here.

Reply:
I changed to bitstream (FieldBool).